### PR TITLE
Setup server to listen on any local address

### LINKED ISSUE #42: restraintd fails to start if loopback interface doesn't support both, IPv6 and IPv4
If the loopback interface supports both, IPv4 and IPv6, `restraintd` starts and binds to both addresses:
```
$ ip add
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever

$ ./restraintd &
Listening on http://localhost:42291

$ netstat -tnap

Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
tcp        0      0 127.0.0.1:42291         0.0.0.0:*               LISTEN      20972/./restraintd          
tcp6       0      0 ::1:42291               :::*                    LISTEN      20972/./restraintd  

```
If IPv6 is not available, `restraintd` fails to start:
```
$ sysctl -w net.ipv6.conf.lo.disable_ipv6=1
net.ipv6.conf.lo.disable_ipv6 = 1

$ ip add
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever

$ ./restraintd &

** (restraintd:20940): ERROR **: 09:48:17.006: Unable to listen on either ipV4 or ipV6 protocols, exiting...

Trace/breakpoint trap (core dumped)
```
The same issue appears if only IPv6 is available on the interface.

`restraintd` should bind to any address available on the loopback interface, IPv4 or IPv6, and fail if there is none.

Originally reported in https://bugzilla.redhat.com/show_bug.cgi?id=1825548 by @jvillal-amp

### PR DESCRIPTION
**Pull request by danrodrig**
The soup_server_listen_local() function tries to listen on both local
addresses, IPv4 and IPv6, and only succeeds when both are bind, even
when the loopback interface doesn't support one of the protocols.

Add rstrnt_listen_any_local() which succeeds when at least one local
address is bind.

Closes #42

### CODE REVIEW

File: src/server.c [hunk 1]
```diff
@@ -574,6 +574,54 @@ null_log_writer (GLogLevelFlags   log_level,
   return G_LOG_WRITER_HANDLED;
 }
 
+/* Bind server to available IPv4 and IPv6 local addresses
+ *
+ * The server must be NOT listening on any interface.
+ *
+ * If port is 0, the server will find an unsued port to listen on. The
```
Comment (JohnVillalovos) [inline src/server.c:581]: nit:  s/unsued/unused/

File: src/server.c [hunk 2]
```diff
@@ -574,6 +574,54 @@ null_log_writer (GLogLevelFlags   log_level,
   return G_LOG_WRITER_HANDLED;
 }
 
+/* Bind server to available IPv4 and IPv6 local addresses
+ *
+ * The server must be NOT listening on any interface.
```
Comment (JohnVillalovos) [inline src/server.c:579]: What does this mean? Does this mean the server must not already be listening on any interface? If so, I suggest this comment:

  The server must NOT already be listening on any interface

File: src/server.c (hunk 1 above)
Comment (danrodrig) [reply src/server.c:581]: Indeed. No ports were sued in this PR...

File: src/server.c (hunk 2 above)
Comment (danrodrig) [reply src/server.c:579]: > Does this mean the server must not already be listening on any interface?

Exactly that. I will update the comment. Thanks for the suggestion!

File: src/server.c [hunk 3]
```diff
@@ -574,6 +574,54 @@ null_log_writer (GLogLevelFlags   log_level,
...
+ * If port is 0, the server will find an unsued port to listen on. The
+ * port will be the same for both addresses.
+ *
+ * Returns TRUE if the server was binded to at least one address. FALSE
+ * otherwise.
+ */
+static gboolean
+rstrnt_listen_any_local (SoupServer *server, guint port)
+{
+    GError   *error;
```
Comment (cbouchar) [inline src/server.c:590]: Should initialize to NULL;

File: src/server.c [hunk 4]
```diff
@@ -574,6 +574,54 @@ null_log_writer (GLogLevelFlags   log_level,
...
+    /* Ensure that server is not listening on any interface */
+    uris = soup_server_get_uris (server);
+    is_listening = g_slist_length (uris) > 0;
+    g_slist_free (uris);
+    g_return_val_if_fail (!is_listening, FALSE);
+
+    is_listening = soup_server_listen_local (server, port, SOUP_SERVER_LISTEN_IPV4_ONLY, &error);
+
+    if (!is_listening) {
+        g_warning ("Unable to listen on local IPv4 address: %s\n", error->message);
```
Comment (cbouchar) [inline src/server.c:603]: Should test if error it not NULL before using.

File: src/server.c [hunk 5]
```diff
@@ -574,6 +574,54 @@ null_log_writer (GLogLevelFlags   log_level,
...
 }
 
+/* Bind server to available IPv4 and IPv6 local addresses
+ *
+ * The server must be NOT listening on any interface.
+ *
+ * If port is 0, the server will find an unsued port to listen on. The
+ * port will be the same for both addresses.
+ *
+ * Returns TRUE if the server was binded to at least one address. FALSE
```
Comment (cbouchar) [inline src/server.c:584]: when you say "was binded" does that mean it is no longer bound.  I suggest changing this to "is bound".

File: src/server.c (hunk 3 above)
Comment (danrodrig) [reply src/server.c:590]: Yep, I missed this. Thanks for point it out.

File: src/server.c (hunk 5 above)
Comment (danrodrig) [reply src/server.c]: True, corrected.

File: src/server.c (hunk 4 above)
Comment (danrodrig) [reply src/server.c:605]: soup_server_listen_local sets error when FALSE is returned, but I agree is better to check the error instead.

**Comment by danrodrig** [comment]
:roll_eyes:

**Pull request merged**